atlasaction/action.go

```diff
@@ -908,6 +905,113 @@ func (a *Actions) SchemaApply(ctx context.Context) error {
+				p.URL = a.GetInput("url")
+				p.Format = `{{ .Hash | base64url }}`
+			},
+		))
+		if err != nil {
+			return nil, fmt.Errorf("failed to inspect the schema: %w", err)
+		}
+		// Create a new schema plan.
+		plan, err := a.Atlas.SchemaPlan(ctx, a.schemaPlanParams(
+			func(p *atlasexec.SchemaPlanParams) {
+				p.From = []string{a.GetInput("url")}
+				p.To = []string{a.GetInput("to")}
+				p.Repo = repo
+				p.Name = fmt.Sprintf("atlas-action-plan-%.8s", strings.ToLower(tag))
+				p.Pending = true
+			},
+		))
+		if err != nil {
+			return nil, fmt.Errorf("failed to create schema plan: %w", err)
+		}
+		// Wait for the plan to be approved.
+		if err := a.waitingForApproval(func() (bool, error) {
+			plans, err := a.Atlas.SchemaPlanList(ctx, params)
+			if err != nil {
+				return false, err
+			}
+			// Check the created plan is exists and approved.
+			if len(plans) == 0 || (len(plans) == 1 && plans[0].Name != plan.File.Name) {
+				return false, errors.New("schema plan not found")
+			}
+			if len(plans) == 1 && plans[0].Status == "APPROVED" {
+				return true, nil
+			}
+			a.Warningf("Schema plan is pending approval, review here: %s", plan.File.Link)
+			return false, nil
+		}); err != nil {
+			if err == ErrTimeoutExceeded {
+				return nil, fmt.Errorf(
+					`the schema plan %q was not approved within the specified waiting period. Please manually approve the plan and re-run the action.
+You can approve the plan by visiting: %s`, plan.File.Name, plan.File.Link)
+			}
+			return nil, err
+		}
+		return a.Atlas.SchemaApplySlice(ctx, a.schemaApplyParams(
+			func(p *atlasexec.SchemaApplyParams) {
+				p.PlanURL = plan.File.URL
+			},
+		))
+	}
+	// Check existing plans and decide to create a new plan.
+	genPlan := a.GetInput("auto-generate-plan")
+	switch plans, err := a.Atlas.SchemaPlanList(ctx, params); {
+	case err != nil:
+		return nil, fmt.Errorf("failed to list schema plans: %w", err)
+	// There are multiple pending plans. This is an unexpected state.
+	case len(plans) > 1:
+		return nil, errors.New("found multiple schema pending plans, please approve or delete the existing plans")
+	// There are no pending plans or the policy is set to "ALWAYS". Create a new plan.
+	case len(plans) == 0 && genPlan == "ALWAYS":
+		return createAndApprovePlan()
+	// There are no pending plans and generate plan option is set to "REVIEW".
+	// In this case, we need to apply the schema. If there are any errors or warnings
+	// caused by the lint policy, we will create an approval plan.
+	case len(plans) == 0 && genPlan == "REVIEW":
+		results, err := a.Atlas.SchemaApplySlice(ctx, a.schemaApplyParams())
+		if err != nil && strings.HasPrefix(err.Error(), "Rejected by review policy") {
+			return createAndApprovePlan()
+		}
+		return results, err
+	case len(plans) == 1:
+		return a.Atlas.SchemaApplySlice(ctx, a.schemaApplyParams(
+			func(p *atlasexec.SchemaApplyParams) {
+				p.PlanURL = plans[0].URL
+			},
+		))
```

Comment:
Atlas should auto-detect this, and I think you don't really need to pass this explicitly.

Reply:
We should keep this one, this is making code easy to read and avoid hidden login in atlas-cli

Reply:
Hi @a8m, any objections?